lib/icingadb/icingadb-worker.cpp

```diff
@@ -0,0 +1,509 @@
+{
+	if (GetActive() && m_Rcon && m_Rcon->IsConnected()) {
+		{
+			std::lock_guard lock(m_PendingItemsMutex);
+			m_PendingItems.insert(PendingDependencyItem{depGroup, child, false, DependencyChildRegistered});
+		}
+		m_PendingItemsCV.notify_one();
+	}
+}
+
+/**
+ * Enqueue the removal of a dependency child from a dependency group.
+ *
+ * This function handles the removal of a child checkable from a dependency group by first checking if there
+ * are any pending registration items for the same child and dependency group. If such an item exists, it is
+ * removed from the pending items queue, effectively canceling the registration. If there is also a pending
+ * dependency group state update triggered by the same child, it is either removed or updated to use a different
+ * child if the group is not being removed entirely. If no pending registration exists, the function proceeds
+ * to enqueue the necessary deletions in Redis for the dependencies and related nodes and edges.
+ *
+ * @param depGroup The dependency group from which the child is being removed.
+ * @param dependencies The list of dependencies associated with the child being removed.
+ * @param removeGroup A flag indicating whether the entire dependency group should be removed.
+ */
+void IcingaDB::EnqueueDependencyChildRemoved(
+	const DependencyGroup::Ptr& depGroup,
+	const std::vector<Dependency::Ptr>& dependencies,
+	bool removeGroup
+)
+{
+	if (dependencies.empty() || !GetActive() || !m_Rcon || !m_Rcon->IsConnected()) {
+		return; // No need to enqueue anything if we're not connected or there are no dependencies.
+	}
+
+	Checkable::Ptr child(dependencies.front()->GetChild());
+	bool hadPendingRegistration = false; // Whether we had a pending child registration to cancel.
+
+	{
+		std::lock_guard lock(m_PendingItemsMutex);
+		if (auto it(m_PendingItems.find(std::make_pair(child, depGroup))); it != m_PendingItems.end()) {
+			hadPendingRegistration = true;
+			m_PendingItems.erase(it); // Cancel the pending child registration.
+			if (auto sit = m_PendingItems.find(std::make_pair(nullptr, depGroup)); sit != m_PendingItems.end()) {
+				if (removeGroup) {
+					m_PendingItems.erase(sit);
+				} else {
+					m_PendingItems.modify(sit, [&child, &depGroup](PendingItemVariant& item) {
+						if (auto& pdit = std::get<PendingDependencyItem>(item); pdit.Child == child) {
+							// If the pending dependency group state update is triggered by the same child being
+							// removed, we can't just cancel it, as this update might have been also requested by
+							// other checkables but since we only track the first one that requested the update, we
+							// need to pick a random other child to trigger the update instead. It's safe to do so,
+							// because the child pointer doesn't contribute to the uniqueness of the pending item.
+							pdit.Child = depGroup->GetAnyChild();
```

Comment:
Instead of tracking a child and updating it, wouldn't it make more sense to just call `GetAnyChild()` when processing the entry when it's actually needed?

Reply:
Done.